### ISSUE TITLE
C++11 compatible?

### ISSUE THREAD

**Issue by falkTX** [sfztools/sfizz#110]
This is a general question. Is there interest to keep code C++11 compatible?

Otherwise it makes it difficult to get the plugin into the MOD units.
In MOD we are still using gcc 4.9, as it can build up to c++11 and that has been enough so far for all the plugins.

**Comment by paulfd** [comment]
Hi! Yes! we have compatibility libraries in C++11 for filesystem, string views and char conversion, which are the key elements we use in C++17.  We spent some time working on compatibility some months ago, I'm not sure we went as low as gcc 4.9 though.

**Comment by falkTX** [comment]
Oh, good to know :)

I have to update cmake to 3.5 still, but that is much easier than updating the compiler.

Currently with master branch I have to following errors:
```
src/sfizz/Defaults.h:71:47: error: 'constexpr sfz::Range<Type>::Range(Type, Type) [with Type = int]' called in a constant expression
  constexpr Range<int> bendRange { -8192, 8192 };
                                               ^
src/sfizz/Range.h:40:15: note: 'constexpr sfz::Range<Type>::Range(Type, Type) [with Type = int]' is not usable as a constexpr function because:
     constexpr Range(Type start, Type end) noexcept
               ^
src/sfizz/SfzHelpers.h:202:16: error: invalid return type 'void' of constexpr function 'constexpr void sfz::multiplyByCents(float&, int)'
 constexpr void multiplyByCents(float& base, int modifier)
                ^
src/sfizz/Opcode.h: In function 'absl::optional<bool> sfz::readBooleanFromOpcode(const sfz::Opcode&)':
src/sfizz/Opcode.h:98:20: error: 'constexpr uint64_t hash(absl::string_view, uint64_t)' called in a constant expression
     case hash("off"):
                    ^
```

**Comment by paulfd** [comment]
The void one is gcc 5 apparently, a bit sad but well. The last one may be because of the `absl::string_view` object, the first one would be constexpr constructor being disallowed?

I seem to remember a MOD Docker environment, should I use this to try and correct these build errors maybe?

**Comment by falkTX** [comment]
You can do it with the mod-plugin-builder, I added a sfizz plugin package in the `sfizz` branch.
The docker stuff is only needed in case you are not running Linux.

I am checking which changes we need to do, to get a confirmation that they are okay with you.
I am fine with sending you patches, no worries on that.

**Comment by jpcima** [comment]
You can reproduce this just by setting CMAKE_CXX_STANDARD to 11.

I remember some restrictions on C++11, where constexpr functions can only consist of a single return statement I believe.

**Comment by falkTX** [comment]
Yeah, that is the biggest issue at the moment.
I have to replace a few constexpr functions with inline ones. The `hash` one is specially nasty.

**Comment by paulfd** [comment]
@falkTX having constexpr hashing for string literals in switch functions is the key thing needed from what you posted. The above `string_view` overload of the hash function is something I think I added, which probably does not need to be `constexpr`. If you call `hash("string literal")` I would guess the `constexpr uint64_t hash(const char*, ...)` overload would be called? This one should be C++11?

**Comment by paulfd** [comment]
Sorry, checking the source, I realize there is no such overload because the `string_view` one handles it. Hopefully adding the overload will be enough. If you open a PR with the rest I can work on this part later today if you have no time :slightly_smiling_face:

**Comment by falkTX** [comment]
I am not sure I follow with the method that needs to be overloaded..

**Comment by jpcima** [comment]
None, because constexpr string_view works in C++11. It's not the root of the problem.

It's just needed to make the function have a single return statement, by rewriting hash to use ternary conditionals.

**Comment by jpcima** [comment]
See: https://godbolt.org/z/TtNPHK

**Comment by falkTX** [comment]
that seems to make it go further, but the switch case fails.
so I have this:
```
constexpr uint64_t hash(absl::string_view s, uint64_t h = Fnv1aBasis)
{
    return s.length() > 0 ? hash( { s.data() + 1, s.length() - 1 }, (h ^ s.front()) * Fnv1aPrime ) : 0;
}
```
but that fails with:
```
src/sfizz/Opcode.h: In function 'absl::optional<bool> sfz::readBooleanFromOpcode(const sfz::Opcode&)':
src/sfizz/Opcode.h:100:5: error: duplicate case value
     case hash("on"):
     ^
src/sfizz/Opcode.h:98:5: error: previously used here
     case hash("off"):
     ^
```

**Comment by paulfd** [comment]
You have to put `h` instead of `0` at the end of the ternary.

We may all be doing the same thing in parallel so I suggest to open a PR and work it out there. Note that it seems the `atomic_queue` used in the background loader may need work too.

**Comment by falkTX** [comment]
ah yes!

so far I am only seeing how far I can make it build

**Comment by falkTX** [comment]
btw, gcc4 does not have any constexpr on the std math functions, so a lot of stuff has to be use `inline` instead of `constexpr`

**Comment by paulfd** [comment]
I made a `constexpr` min/max using a ternary too, it works out and that should be the only constexpr thing needed for maths (even then it's just a sanity check on the constexpr `Range` which is not really critical...).

**Comment by paulfd** [comment]
This rule about only a single return statement does not encourage readable code lol, I'm checking out what will be needed in the `atomic_queue` and it will be ugly :D

**Comment by falkTX** [comment]
The Range constexpr I was able to get around by using this:
```
    constexpr Range(Type start, Type end) noexcept
        : _start(start)
        , _end(start > end ? start : end)
    {
    }
```

**Comment by paulfd** [comment]
Ho no..

```
/home/paul/source/sfizz/src/sfizz/Synth.cpp:88:35: error: use of ‘auto’ in lambda parameter declaration only available with ‘-std=c++14’ or ‘-std=gnu++14’
```
This will sting xD

**Comment by falkTX** [comment]
These are the changes I made so far.
```
diff --git a/src/sfizz/Buffer.h b/src/sfizz/Buffer.h
index 757fe9c..e37b634 100644
--- a/src/sfizz/Buffer.h
+++ b/src/sfizz/Buffer.h
@@ -274,9 +274,9 @@ public:
         return counter;
     }
 private:
-    static constexpr auto AlignmentMask { Alignment - 1 };
-    static constexpr auto TypeAlignment { Alignment / sizeof(value_type) };
-    static constexpr auto TypeAlignmentMask { TypeAlignment - 1 };
+    static constexpr auto AlignmentMask = Alignment - 1U;
+    static constexpr auto TypeAlignment = Alignment / sizeof(value_type);
+    static constexpr auto TypeAlignmentMask = TypeAlignment - 1U;
     static_assert(std::is_arithmetic<value_type>::value, "Type should be arithmetic");
     static_assert(Alignment == 0 || Alignment == 4 || Alignment == 8 || Alignment == 16, "Bad alignment value");
     static_assert(TypeAlignment * sizeof(value_type) == Alignment, "The alignment does not appear to be divided by the size of the Type");
diff --git a/src/sfizz/EGDescription.h b/src/sfizz/EGDescription.h
index cc65551..ceba155 100644
--- a/src/sfizz/EGDescription.h
+++ b/src/sfizz/EGDescription.h
@@ -134,9 +134,12 @@ struct EGDescription
      * @param velocity
      * @return float
      */
-    float getStart(const SfzCCArray &ccValues, uint8_t velocity [[maybe_unused]]) const noexcept
+    float getStart(const SfzCCArray &ccValues, uint8_t velocity) const noexcept
     {
         return Default::egPercentRange.clamp(ccSwitchedValue(ccValues, ccStart, start));
+
+        // unused
+        (void)velocity;
     }
     /**
      * @brief Get the sustain level with possibly a CC modifier and a velocity modifier
diff --git a/src/sfizz/MathHelpers.h b/src/sfizz/MathHelpers.h
index aa99bec..60aec45 100644
--- a/src/sfizz/MathHelpers.h
+++ b/src/sfizz/MathHelpers.h
@@ -127,9 +127,7 @@ inline float midiNoteFrequency(const int noteNumber)
 template<class T>
 constexpr T clamp( T v, T lo, T hi )
 {
-    v = std::min(v, hi);
-    v = std::max(v, lo);
-    return v;
+    return std::max(std::min(v, hi), lo);
 }
 
 template<int Increment = 1, class T>
@@ -152,17 +150,17 @@ constexpr ValueType linearInterpolation(ValueType left, ValueType right, ValueTy
 }
 
 template <class Type>
-constexpr Type pi { static_cast<Type>(3.141592653589793238462643383279502884) };
+constexpr Type pi() { return static_cast<Type>(3.141592653589793238462643383279502884); };
 template <class Type>
-constexpr Type twoPi { static_cast<Type>(2) * pi<Type> };
+constexpr Type twoPi() { return static_cast<Type>(2) * pi<Type>(); };
 template <class Type>
-constexpr Type piTwo { pi<Type> / static_cast<Type>(2) };
+constexpr Type piTwo() { return pi<Type>() / static_cast<Type>(2); };
 template <class Type>
-constexpr Type piFour { pi<Type> / static_cast<Type>(4) };
+constexpr Type piFour() { return pi<Type>() / static_cast<Type>(4); };
 template <class Type>
-constexpr Type sqrtTwo { static_cast<Type>(1.414213562373095048801688724209698078569671875376948073176) };
+constexpr Type sqrtTwo() { return static_cast<Type>(1.414213562373095048801688724209698078569671875376948073176); };
 template <class Type>
-constexpr Type sqrtTwoInv { static_cast<Type>(0.707106781186547524400844362104849039284835937688474036588) };
+constexpr Type sqrtTwoInv() { return static_cast<Type>(0.707106781186547524400844362104849039284835937688474036588); };
 
 /**
    @brief A fraction which is parameterized by integer type
diff --git a/src/sfizz/MidiState.h b/src/sfizz/MidiState.h
index afba37c..e48e776 100644
--- a/src/sfizz/MidiState.h
+++ b/src/sfizz/MidiState.h
@@ -134,13 +134,13 @@ private:
      * @brief Stores the note on times.
      *
      */
-	MidiNoteArray<NoteOnTime> noteOnTimes { };
+	MidiNoteArray<NoteOnTime> noteOnTimes;
     /**
      * @brief Stores the velocity of the note ons for currently
      * depressed notes.
      *
      */
-	MidiNoteArray<uint8_t> lastNoteVelocities { };
+	MidiNoteArray<uint8_t> lastNoteVelocities;
     /**
      * @brief Current known values for the CCs.
      *
diff --git a/src/sfizz/Range.h b/src/sfizz/Range.h
index acb88c6..27fec5d 100644
--- a/src/sfizz/Range.h
+++ b/src/sfizz/Range.h
@@ -39,7 +39,7 @@ public:
     // }
     constexpr Range(Type start, Type end) noexcept
         : _start(start)
-        , _end(std::max(start, end))
+        , _end(start > end ? start : end)
     {
     }
     ~Range() = default;
diff --git a/src/sfizz/SIMDHelpers.h b/src/sfizz/SIMDHelpers.h
index f4c5a2d..678a985 100644
--- a/src/sfizz/SIMDHelpers.h
+++ b/src/sfizz/SIMDHelpers.h
@@ -734,7 +734,7 @@ namespace _internals {
         int i = 0;
 
         for (; i < panSize; ++i)
-            pan[i] = std::cos(i * (piTwo<double> / (panSize - 1)));
+            pan[i] = std::cos(i * (piTwo<double>() / (panSize - 1)));
 
         for (; i < static_cast<int>(pan.size()); ++i)
             pan[i] = pan[panSize - 1];
diff --git a/src/sfizz/SfzHelpers.h b/src/sfizz/SfzHelpers.h
index 1e590d1..fb8ce05 100644
--- a/src/sfizz/SfzHelpers.h
+++ b/src/sfizz/SfzHelpers.h
@@ -81,7 +81,7 @@ constexpr float normalizePercents(T percentValue)
  * @param bendValue
  * @return constexpr float
  */
-constexpr float normalizeBend(float bendValue)
+inline float normalizeBend(float bendValue)
 {
     return std::min(std::max(bendValue, -8191.0f), 8191.0f) / 8191.0f;
 }
@@ -199,7 +199,7 @@ constexpr void addToBase(T& base, T modifier)
  * @param base
  * @param modifier
  */
-constexpr void multiplyByCents(float& base, int modifier)
+inline void multiplyByCents(float& base, int modifier)
 {
     base *= centsFactor(modifier);
 }
diff --git a/src/sfizz/StringViewHelpers.h b/src/sfizz/StringViewHelpers.h
index 0cc49d3..7e198e8 100644
--- a/src/sfizz/StringViewHelpers.h
+++ b/src/sfizz/StringViewHelpers.h
@@ -61,8 +61,5 @@ constexpr uint64_t Fnv1aPrime = 0x01000193;
  */
 constexpr uint64_t hash(absl::string_view s, uint64_t h = Fnv1aBasis)
 {
-    if (s.length() > 0)
-        return hash( { s.data() + 1, s.length() - 1 }, (h ^ s.front()) * Fnv1aPrime );
-
-    return h;
+    return s.empty() ? h : hash( { s.data() + 1, s.length() - 1 }, (h ^ s.front()) * Fnv1aPrime );
 }
diff --git a/src/sfizz/Synth.cpp b/src/sfizz/Synth.cpp
index 0c3b3f8..cf024a5 100644
--- a/src/sfizz/Synth.cpp
+++ b/src/sfizz/Synth.cpp
@@ -156,7 +156,7 @@ void sfz::Synth::handleControlOpcodes(const std::vector<Opcode>& members)
         const auto backParameter = member.backParameter();
         switch (member.lettersOnlyHash) {
         case hash("Set_cc"):
-            [[fallthrough]];
+            // fall-through
         case hash("set_cc"):
             if (backParameter && Default::ccNumberRange.containsWithEnd(*backParameter)) {
                 const auto ccValue = readOpcode(member.value, Default::ccValueRange).value_or(0);
@@ -164,13 +164,13 @@ void sfz::Synth::handleControlOpcodes(const std::vector<Opcode>& members)
             }
             break;
         case hash("Label_cc"):
-            [[fallthrough]];
+            // fall-through
         case hash("label_cc"):
             if (backParameter && Default::ccNumberRange.containsWithEnd(*backParameter))
                 ccNames.emplace_back(*backParameter, std::string(member.value));
             break;
         case hash("Default_path"):
-            [[fallthrough]];
+            // fall-through
         case hash("default_path"):
             defaultPath = absl::StrReplaceAll(trim(member.value), { { "\\", "/" } });
             DBG("Changing default sample path to " << defaultPath);
@@ -273,7 +273,7 @@ bool sfz::Synth::loadSfzFile(const fs::path& file)
                 region->isStereo = true;
 
             // TODO: adjust with LFO targets
-            const auto maxOffset { region->offset + region->offsetRandom };
+            const uint maxOffset { region->offset + region->offsetRandom };
             if (!resources.filePool.preloadFile(region->sample, maxOffset))
                 removeCurrentRegion();
         }
@@ -363,7 +363,7 @@ sfz::Voice* sfz::Synth::findFreeVoice() noexcept
 
 int sfz::Synth::getNumActiveVoices() const noexcept
 {
-    auto activeVoices { 0 };
+    int activeVoices { 0 };
     for (const auto& voice : voices) {
         if (!voice->isFree())
             activeVoices++;
@@ -465,7 +465,7 @@ void sfz::Synth::noteOn(int delay, int noteNumber, uint8_t velocity) noexcept
     noteOnDispatch(delay, noteNumber, velocity);
 }
 
-void sfz::Synth::noteOff(int delay, int noteNumber, uint8_t velocity [[maybe_unused]]) noexcept
+void sfz::Synth::noteOff(int delay, int noteNumber, uint8_t velocity) noexcept
 {
     ASSERT(noteNumber < 128);
     ASSERT(noteNumber >= 0);
@@ -485,7 +485,10 @@ void sfz::Synth::noteOff(int delay, int noteNumber, uint8_t velocity [[maybe_unu
     for (auto& voice : voices)
         voice->registerNoteOff(delay, noteNumber, replacedVelocity);
 
-    noteOffDispatch(delay, noteNumber, replacedVelocity);
+    return noteOffDispatch(delay, noteNumber, replacedVelocity);
+
+    // unused
+    (void)velocity;
 }
 
 void sfz::Synth::noteOffDispatch(int delay, int noteNumber, uint8_t velocity) noexcept
```
I will stop here since you seem to be doing the same :)

The warnings can be ignored, like unknown [[maybe_unused]] stuff, I just did it locally here so I could better see results.

the `atomic_queue::round_up_to_power_of_2` really seems like going to be the biggest pain

**Comment by paulfd** [comment]
Yes we'll need to patch this atomic_queue... Ho well. I'll work on updating all of this. Thanks for the report and the patches!

We might also need a gcc 4.9 CI run to check that we're not introducing problems.

**Comment by jpcima** [comment]
I'll suggest to rather a MOD plugin builder CI, based on the Docker image.

**Comment by paulfd** [comment]
Changes have been merged in #111 and we have a CI check in #112 , so I'm closing this for now. I'll let you open other issues as needed! Thanks.
